discovery.go

```diff
@@ -0,0 +1,310 @@
+package pubsub
+
+import (
+	"context"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+var (
+	// poll interval
+	DiscoveryPollInitialDelay = 0 * time.Millisecond
+	DiscoveryPollInterval     = 1 * time.Second
+)
+
+// discover represents the discovery pipeline.
+// The discovery pipeline handles advertising and discovery of peers
+type discover struct {
+	p *PubSub
+
+	// discovery assists in discovering and advertising peers for a topic
+	discovery discovery.Discovery
+
+	// advertising tracks which topics are being advertised
+	advertising map[string]context.CancelFunc
+
+	// discoverQ handles continuing peer discovery
+	discoverQ chan *discoverReq
+
+	// ongoing tracks ongoing discovery requests
+	ongoing map[string]struct{}
+
+	// done handles completion of a discovery request
+	done chan string
+}
+
+// MinTopicSize returns a function that checks if a router is ready for publishing based on the topic size.
+// The router ultimately decides the whether it is ready or not, the given size is just a suggestion.
+func MinTopicSize(size int) RouterReady {
+	return func(rt PubSubRouter, topic string) (bool, error) {
+		return rt.EnoughPeers(topic, size), nil
+	}
+}
+
+// newDiscover creates a new peer discovery mechanism
+func newDiscover() *discover {
+	return &discover{
+		advertising: make(map[string]context.CancelFunc),
+		discoverQ:   make(chan *discoverReq, 32),
+		ongoing:     make(map[string]struct{}),
+		done:        make(chan string),
+	}
+}
+
+// Start attaches the discovery pipeline to a pubsub instance and starts event loop
+func (d *discover) Start(p *PubSub) {
+	if d.discovery == nil || p == nil {
+		return
+	}
+
+	d.p = p
+	go d.discoverLoop()
+	go d.pollTimer()
+}
+
+func (d *discover) pollTimer() {
+	select {
+	case <-time.After(DiscoveryPollInitialDelay):
+	case <-d.p.ctx.Done():
+	}
```

Comment:
ya, might as well.

Reply:
done. Also, when you make comments in the future could you comment on the files tab instead of on a particular iteration. For some reason your comments don't show "outdated" after an update.